### ISSUE TITLE
[@typescript-eslint/scope-manager] The property of a non-computed MemberExpression node is not referenced

### ISSUE THREAD

**Issue by testmynamelist** [typescript-eslint/typescript-eslint#2534]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I'm moving my issue from [[Question] How to find reference for a class property used in a non-computed MemberExpression node · Issue #65 · eslint/eslint-scope](https://github.com/eslint/eslint-scope/issues/65) here, if I understand right, this is an experimental utils one, please correct me if I'm opening the issue in wrong type.

I'm developing a custom rule for typescript, and one requirement is to collect the references for a variable or class property, but I failed to get the expected reference map in one scenario which is detailed below. After checking the `referencer.ts` source code, it looks to me like a limitation, thus raised this issue for your help, thank you

**Repro**

The .ts source file to be linted:
```TS
export class Hi {
  private aProperty = 'Test';
  private getAProperty() {
    this.generate({
      prop: this.aProperty
    });
  }
}
```

and my custom rule is using below code to try to collect references, aiming to find that class property `aProperty` is referenced by Property value node `this.aProperty`

```
 const scope = context.getScope();

    function createReferenceMap(scope, outReferenceMap = new Map()) {
      for (let reference of scope.references) {
        outReferenceMap.set(reference.identifier, reference);
      }
      for (const childScope of scope.childScopes) {
        createReferenceMap(childScope, outReferenceMap);
      }

      return outReferenceMap;
    }
```

**Expected Result**

Expecting class property `aProperty` is referenced by Property value node `this.aProperty`

**Actual Result**

but unfortunately, `this.aProperty` is not listed in the reference map after calling the function `createReferenceMap`, the `MemberExpression` node which has `computed` attr as `false`. 

**Additional Info**

After debugging, I found below codes in the [referencer](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/scope-manager/src/referencer/Referencer.ts), and found it's the node.computed condition prevents the above class property being referenced.

```
    MemberExpression(node) {
        this.visit(node.object);
        if (node.computed) {
            this.visit(node.property);
        }
    }
```
**Versions**

| package                                 | version |
| --------------------------------------- | ------- |
| `@typescript-eslint/experimental-utils` | `X.Y.Z` |
| `TypeScript`                            | `X.Y.Z` |
| `node`                                  | `10.16.0` |

**Comment by bradzacher** [comment]
This is entirely intentional.

You're misunderstanding how we track computed property references. We do not track the reference to the property itself - we only track the reference to the variable. For `this[computed]` we track the reference to `this`, and the reference to `computed`. This does not take into account the _value_ of `computed`, just the fact that it's a variable that's being referenced (i.e if `computed === 'foo'`, we will not track a reference to `this.foo`).

What you're doing is referencing a property (`aProperty`) on a variable (`this`).
ESLint's scope analysis does not (and never has) tracked _property_ references.

Classes are no exception to the this.

We cannot reliably track these references for the following reasons, so instead of providing incomplete and inconsistent tracking, we just do not do any tracking.

- In TS you may declare properties on the class itself, but there is no requirement to do so - there are the following cases which make it.
  - For example, you may `@ts-ignore` the undefined property error, 
- you may declare properties on a super-class, which means we cannot track references to that property
```ts
class Foo { a: string }
class Bar extends Foo {
  method() {
    this.a // we would not be able to resolve this correctly
  }
}
```
- you may declare properties via an interface declaration merge, which means we cannot track references to that property.
```ts
interface Foo { a: string }
class Foo {
  method() {
    this.a // we would not be able to resolve this correctly
  }
}
```
- outside of the class, we have no ability to track references to class properties.
```ts
class Foo { a: string }
const x = new Foo();
x.a // we would not be able to resolve this correctly
```
- There would also exist a problem in that we cannot tell for certain what the `this` references.
```ts
function foo() {
  this.a  // we would not be able to resolve this correctly
}
class Foo {
  method = foo;
}
```

I could go on.

This is not a feature that will be added to the scope analyser.

**Comment by testmynamelist** [comment]
@bradzacher thank you very much for the detailed explanation! I understand now, we may need to seek other ways to link the class property `aProperty` with its usage `this.aProperty`, do you see if there is any possible way we can do this without referencer?

**Comment by bradzacher** [comment]
The scope analyser will not be able to help you do this at all.

You will instead need to just manually do ast analysis and look for the this expressions.
